### PR TITLE
fix: reduce float and half-float values to their stored precision

### PR DESCRIPTION
**Pull request by salvatore-campagna**
This is a fix for #82867 and #81529.

The bug is a result of reducing precision of `float` values. If the filed the range aggregation is applied to is a `long` field we try to reduce the precision of the rage `from` and/or `to` values to `long`. This results in an exception at parsing time happening when parsing a `long` with decimal positions.

Here we parse `FLOAT` and `HALF-FLOAT` values actually reducing their precision. For other types, including `long` we do nothing, or better, we do not do any "precision-reduction" action in method `reduceToStoredPrecision`.

### CODE REVIEW

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/40_range.yml [hunk 1]
```diff
@@ -199,6 +199,45 @@ setup:
   - is_false:  aggregations.float_range.buckets.2.to
   - match: { aggregations.float_range.buckets.2.doc_count: 3 }
 
+---
+"Double range on long field":
+  - skip:
+      version: " - 7.16.99"
```
Comment (nik9000) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/40_range.yml:205]: I think this has to be 8.0.99 or the BWC tests will fail while you backport.

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/docs

**Review by not-napoleon — APPROVED** [review]
LGTM.  Thanks for jumping on this.

**Comment by salvatore-campagna** [comment]
The bug is a result of reducing precision of float values. If the filed the range aggregation is applied to is a long field, we try to reduce the precision of the rage `from` and/or `to` values to `long`. This results in an exception at parsing time happening when parsing a long with decimal positions.

Here we parse FLOAT and HALF-FLOAT values actually reducing their precision. For other types, including `long` we do nothing, or better, we do not do any "precision-reduction" action in method reduceToStoredPrecision.

**Comment by salvatore-campagna** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.0|[](https://github.com/elastic/elasticsearch/pull/83410)|
|✅|7.17|[](https://github.com/elastic/elasticsearch/pull/83411)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by salvatore-campagna** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.0|[](https://github.com/elastic/elasticsearch/pull/83415)|
|✅|7.17|[](https://github.com/elastic/elasticsearch/pull/83416)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by salvatore-campagna** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.0|[](https://github.com/elastic/elasticsearch/pull/83415)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by JWilsonElastic** [comment]
Just swinging by to ask if this backport ultimately failed. It looks like the backport PR failed, but wanted to confirm there was no successful backport for 7.x.

**Pull request merged**